### PR TITLE
Add GCP service account support to BigQuery

### LINKED ISSUE #4634: BigQuery should use Service Accounts for OAuth
Instead of an admin [copy/pasting an auth code tied to their Google account](http://www.metabase.com/docs/v0.23.0/administration-guide/databases/bigquery.html), Metabase should support using [Service Accounts](https://developers.google.com/identity/protocols/OAuth2ServiceAccount), which would instead involve uploading a Google-provided JSON file containing a service account address, keypair, and some other metadata (see [Redash's docs](https://redash.io/help/data-sources/bigquery.html) for an idea of what this process is like). The current auth system is bound to cause issues eventually, like when the person who set up Metabase changes jobs and their account is deleted.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by robdaemon**
Adds the ability to use a service account JSON key file for talking to
Google BigQuery

Resolves #4634

[ci bigquery]

### CODE REVIEW

**Review by camsaul — APPROVED** [review]
This works for me, but can you make sure you run the tests with `[ci bigquery]` before merging?

**Comment by robdaemon** [comment]
@kdoh - this is ready for the FE changes, did you want to add them to this branch / PR?

File: modules/drivers/google/src/metabase/driver/google.clj [hunk 1]
```diff
@@ -70,46 +71,60 @@
...
 
-        ;; don't use `execute` here because this is a *different* type of Google request
+            ;; don't use `execute` here because this is a *different* type of Google request
         ^GoogleTokenResponse response
         (.execute (doto (.newTokenRequest flow auth-code)
                     (.setRedirectUri redirect-uri)))]
     {:access-token (.getAccessToken response), :refresh-token (.getRefreshToken response)}))
 
-(def ^{:arglists '([scopes client-id client-secret auth-code])} fetch-access-and-refresh-tokens
+(def ^{:arglists '([scopes client-id client-secret auth-code service-account-json])} fetch-access-and-refresh-tokens
```
Comment (paulrosenzweig) [inline modules/drivers/google/src/metabase/driver/google.clj:86]: @robdaemon it looks like while `fetch-access-and-refresh-tokens` passes `service-account-json` through to `fetch-access-and-refresh-tokens*`, the unmemoized function doesn't accept that arity.

What should happen? Should `fetch-access-and-refresh-tokens*` accept `service-account-json` but ignore it?

Comment (robdaemon) [reply modules/drivers/google/src/metabase/driver/google.clj:86]: ugh that is totally a bug on my part. `fetch-access-and-refresh-tokens` doesn't need `service-account-json` at all, it won't act on it.

**Comment by paulrosenzweig** [comment]
I ended up removing the connection-properties from the BigQuery driver settings. Implementing a declarative way to express the backwards-compatible form switching logic didn't seem justified here. Instead, all the BQ fields are defined in the front end code. The downside of that is that a third-party driver would be unable to update its form in a backwards compatible way like we did.

I added a "text file" form widget type. It displays a raw browser file picker, but it slurps in the text and POSTs it up with the rest of the data. We also validate that the JSON parses. I added that because non-JSON files get a raw parse error if sent to the server.

Here's what it looks like (cc @kdoh):

### New connection

image

### Existing new-style connection

The file uploader appears blank. I'm not sure how best to represent that while allowing updating.

image

### Existing old-style connection

image

After clicking "Connect to a Service Account instead", it shows the new form but with a link to go back:

image

**Comment by kdoh** [comment]
@paulrosenzweig Looking solid! One thing I'm running into is getting the new style form to validate so I can hit save. After adding a dataset-id and choosing my json file the save button never becomes enabled and there's no feedback on why.

**Comment by sgammon** [comment]
just wanted to report that we're already using this in production and it's fantastic

THANK YOU

**Comment by sgammon** [comment]
i'd love to contribute some future enhancements. if someone wants to @ me on the next PR, or i can get to it in a few weeks, but i can already think of some:

- support for read sessions
- support for fine-grained reader identity (allowing column restrictions)
- query caching integration w/bq
- i don't know there's probably more

but don't let this take away from the current work. fantastic, phenomenal work. it just functions as it should and that's so great

**Comment by kdoh** [comment]
@sgammon Happy to hear it's already helping! We'd definitely like to weave in some more enhancements once the basic support lands so we'll keep you posted.

**Comment by paulrosenzweig** [comment]
@kdoh turns out Safari doesn't have a `.text()` method on files. Just fixed it in a way that MDN says should be broadly compatible.

**Review by kdoh — APPROVED** [review]
Looks good to me!

**Pull request merged**